demos/common/utilities/sample_adu_jws_mbedtls.c

```diff
@@ -0,0 +1,796 @@
+{
+    uint32_t ulIndex = 0;
+
+    while( ulIndex < ulSignatureLength )
+    {
+        if( *pucSignature == '-' )
+        {
+            *pucSignature = '+';
+        }
+        else if( *pucSignature == '_' )
+        {
+            *pucSignature = '/';
+        }
+
+        pucSignature++;
+        ulIndex++;
+    }
+}
+
+/**
+ * @brief Calculate the SHA256 over a buffer of bytes
+ *
+ * @param pucInput The input buffer over which to calculate the SHA256.
+ * @param ulInputLength The length of \p pucInput.
+ * @param pucOutput The output buffer into which the SHA256. It must be 32 bytes in length.
+ * @return uint32_t The result of the operation.
+ * @retval 0 if successful.
+ * @retval Non-0 if not successful.
+ */
+static uint32_t prvJWS_SHA256Calculate( const uint8_t * pucInput,
+                                        uint32_t ulInputLength,
+                                        uint8_t * pucOutput )
+{
+    mbedtls_md_context_t ctx;
+    mbedtls_md_type_t md_type = MBEDTLS_MD_SHA256;
+
+    mbedtls_md_init( &ctx );
+    mbedtls_md_setup( &ctx, mbedtls_md_info_from_type( md_type ), 0 );
+    mbedtls_md_starts( &ctx );
+    mbedtls_md_update( &ctx, pucInput, ulInputLength );
+    mbedtls_md_finish( &ctx, pucOutput );
+    mbedtls_md_free( &ctx );
+
+    return 0;
+}
+
+/**
+ * @brief Verify the manifest via RS256 for the JWS.
+ *
+ * @param pucInput The input over which the RS256 will be verified.
+ * @param ulInputLength The length of \p pucInput.
+ * @param pucSignature The encrypted signature which will be decrypted by \p pucN and \p pucE.
+ * @param ulSignatureLength The length of \p pucSignature.
+ * @param pucN The key's modulus which is used to decrypt \p signature.
+ * @param ulNLength The length of \p pucN.
+ * @param pucE The exponent used for the key.
+ * @param ulELength The length of \p pucE.
+ * @param pucBuffer The buffer used as scratch space to make the calculations. It should be at least
+ * `jwsRSA3072_SIZE` + `jwsSHA256_SIZE` in size.
+ * @param ulBufferLength The length of \p pucBuffer.
+ * @return uint32_t The result of the operation.
+ * @retval 0 if successful.
+ * @retval Non-0 if not successful.
+ */
+static uint32_t prvJWS_RS256Verify( uint8_t * pucInput,
+                                    uint32_t ulInputLength,
+                                    uint8_t * pucSignature,
+                                    uint32_t ulSignatureLength,
+                                    uint8_t * pucN,
+                                    uint32_t ulNLength,
+                                    uint8_t * pucE,
+                                    uint32_t ulELength,
+                                    uint8_t * pucBuffer,
+                                    uint32_t ulBufferLength )
+{
+    AzureIoTResult_t xResult;
+    int32_t lMbedTLSResult;
+    uint8_t * pucShaBuffer;
+    size_t ulDecryptedLength;
+    mbedtls_rsa_context ctx;
+    int shaMatchResult;
+
+    if( ulBufferLength < jwsSHA_CALCULATION_SCRATCH_SIZE )
```

Comment:
Do we need this check?  This is a static function where we should trust caller in general, and in this case specifically `ulBufferLength` the caller has already hard-coded to `jwsSHA_CALCULATION_SCRATCH_SIZE`.

Reply:
not really I guess. Was just to double check. We can just tell people that unexpected behavior will occur if you don't use that macro